basic_system/src/system_implementation/flat_storage_model/storage_tree/mod.rs

```diff
@@ -0,0 +1,2645 @@
+        // we want all "previous index" answers to come from the state of the tree that is before any updates,
+        // so it's convenient for node implementation - we can cache key -> index part before applying an update,
+        // and avoid recomputing updates after every insert. Instead we will carefully relink below
+
+        // NOTE: flat keys come in random order, so we should be extra careful with re-linking
+
+        #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+        enum RelinkData {
+            Deleted {
+                previous_index: u64,
+                deletion_index: u64,
+            },
+            Insert {
+                previous_index: u64,
+                insertion_index: u64,
+            },
+        }
+
+        let mut relinked_slots = BTreeMap::<Bytes32, RelinkData, A>::new_in(allocator.clone());
+
+        for (key, value) in deletes_iter {
+            num_total_writes += 1;
+            let flat_key = derive_flat_storage_key(&key.address, &key.key);
+
+            // TODO: debug implementation for B160 uses global alloc, which panics in ZKsync OS
+            #[cfg(not(target_arch = "riscv32"))]
+            let _ = logger.write_fmt(format_args!(
+                "applying delete for address = {:?}, key = {:?}\n",
+                &key.address, &key.key
+            ));
+            // here we branch because we COULD have requested this one as a read witness before
+
+            let (entry, index) = remove_for_existing_flat_key(
+                &mut key_to_index_cache,
+                &mut index_to_leaf_cache,
+                &flat_key,
+                oracle,
+                saved_next_free_slot,
+            );
+            assert_eq!(entry.current_value(), &value.initial_value);
+            debug_assert!(value.current_value.is_zero());
+            let next_key_to_use = *entry.current_next_key();
+            // we saved "next key", and can now delete it completely
+            entry.mark_deleted();
+
+            let (entry, previous_index) = get_or_insert_previous_for_flat_key(
+                &mut key_to_index_cache,
+                &mut index_to_leaf_cache,
+                &flat_key,
+                oracle,
+                saved_next_free_slot,
+            );
+            assert_ne!(index, previous_index);
+            assert_eq!(entry.current_next_key(), &flat_key);
+
+            // We store a key that has to be re-linked, and two indexes - a slot that previously pointed to it
+            relinked_slots.insert(
+                next_key_to_use,
+                RelinkData::Deleted {
+                    previous_index,
+                    deletion_index: index,
+                },
+            );
+
+            let is_new = empty_slots_cache.insert(index);
+            assert!(is_new);
+        }
+
+        // and last one - insert. We will reuse slots
+        for (key, value) in inserts_iter {
```

Comment:
We could start with inserts and for every insert check if we have something in `deletes_iter`, and if any, consume and process one deletion. It may be a bit more complicated, but this way we don't need to relink slots and can basically override slot